### PR TITLE
TFL: Port HARD_SWISH operator from TFLite to TFLu

### PR DESCRIPTION
**Pull request by mansnils**
This is needed for mobilenet v3.

There is some duplicated code because of references to Gemmlowp. If those could be replaced, the tensorflow lite hard_swish operator could be moved out to its own reference/hard_swish.h that both Lite and Micro could include.

### CODE REVIEW

**Comment by rthadur** [comment]
@mansnils can you please resolve conflicts ?

**Comment by gbaned** [comment]
@mansnils can you please resolve conflicts ?

**Comment by mansnils** [comment]
Conflicts resolved.

**Comment by mansnils** [comment]
Gentle ping for review

**Comment by mansnils** [comment]
@njeffrie ping for review

**Comment by mansnils** [comment]
@njeffrie Can we review this please?

**Review by njeffrie — DISMISSED** [review]

**Review by rthadur — DISMISSED** [review]

**Comment by rthadur** [comment]
@mansnils can you please check below errors 

`/tensorflow/lite/micro/kernels/activations.cc:216:22: error: implicit conversion increases floating-point precision: 'const float' to 'double' 

  QuantizeMultiplier(output_multiplier, &output_multiplier_fixedpoint_int32,
  ~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~
/tensorflow/lite/micro/kernels/activations.cc:225:22: error: implicit conversion increases floating-point precision: 'const float' to 'double' 

  QuantizeMultiplier(reluish_multiplier, &reluish_multiplier_fixedpoint_int32,
  ~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~

/tensorflow/lite/micro/kernels/activations.cc:216:22: error: implicit conversion increases floating-point precision: 'const float' to 'double' 

  QuantizeMultiplier(output_multiplier, &output_multiplier_fixedpoint_int32,
  ~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~

/tensorflow/lite/micro/kernels/activations.cc:330:14: note: in instantiation of function template specialization 'tflite::ops::micro::activations::HardSwishQuantized<unsigned char>' requested here
      return HardSwishQuantized<uint8>(context, node);
             ^

/tensorflow/lite/micro/kernels/activations.cc:225:22: error: implicit conversion increases floating-point precision: 'const float' to 'double'

  QuantizeMultiplier(reluish_multiplier, &reluish_multiplier_fixedpoint_int32,
  ~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~

/tensorflow/lite/micro/kernels/activations.cc:216:22: error: implicit conversion increases floating-point precision: 'const float' to 'double' 

  QuantizeMultiplier(output_multiplier, &output_multiplier_fixedpoint_int32,
  ~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~

/tensorflow/lite/micro/kernels/activations.cc:333:14: note: in instantiation of function template specialization 'tflite::ops::micro::activations::HardSwishQuantized<signed char>' requested here
      return HardSwishQuantized<int8>(context, node);
             ^

/tensorflow/lite/micro/kernels/activations.cc:225:22: error: implicit conversion increases floating-point precision: 'const float' to 'double'

  QuantizeMultiplier(reluish_multiplier, &reluish_multiplier_fixedpoint_int32,
  ~~~~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~~
6 errors generated.`

**Comment by mansnils** [comment]
@rthadur Errors are resolved now.

**Review by njeffrie — DISMISSED** [review]

**Comment by rthadur** [comment]
@mansnils I see this internal errors , please check 

`/tensorflow/lite/micro/kernels/activations_test.cc:175:16: error: variable length array used 
 T output_data[size];
              ^
/tensorflow/lite/micro/kernels/activations_test.cc:176:25: error: variable length array used 
 T input_data_quantized[size];
                       ^
/tensorflow/lite/micro/kernels/activations_test.cc:197:27: error: variable length array used 
 float dequantized_output[output_elements_count];
                         ^
/tensorflow/lite/micro/kernels/activations_test.cc:306:16: error: variable length array used 
 T output_data[size];
              ^
/tensorflow/lite/micro/kernels/activations_test.cc:307:25: error: variable length array used 
 T input_data_quantized[size];
                       ^
/tensorflow/lite/micro/kernels/activations_test.cc:324:27: error: variable length array used 
 float dequantized_output[output_elements_count];
                         ^
/tensorflow/lite/micro/kernels/activations_test.cc:397:20: error: variable length array used 
 float output_data[size];
                  ^
/tensorflow/lite/micro/kernels/activations_test.cc:175:16: error: variable length array used 
 T output_data[size];
              ^
/tensorflow/lite/micro/kernels/activations_test.cc:771:26: note: in instantiation of function template specialization 'tflite::testing::(anonymous namespace)::TestHardSwishQuantized<signed char>' requested here
       tflite::testing::TestHardSwishQuantized<int8_t>(
                        ^
/tensorflow/lite/micro/kernels/activations_test.cc:176:25: error: variable length array used 
 T input_data_quantized[size];
                       ^
/tensorflow/lite/micro/kernels/activations_test.cc:197:27: error: variable length array used 
 float dequantized_output[output_elements_count];
                         ^
/tensorflow/lite/micro/kernels/activations_test.cc:175:16: error: variable length array used 
 T output_data[size];
              ^
/tensorflow/lite/micro/kernels/activations_test.cc:773:26: note: in instantiation of function template specialization 'tflite::testing::(anonymous namespace)::TestHardSwishQuantized<unsigned char>' requested here
       tflite::testing::TestHardSwishQuantized<uint8_t>(
                        ^
/tensorflow/lite/micro/kernels/activations_test.cc:176:25: error: variable length array used 
 T input_data_quantized[size];
                       ^
/tensorflow/lite/micro/kernels/activations_test.cc:197:27: error: variable length array used 
 float dequantized_output[output_elements_count];
                         ^
/tensorflow/lite/micro/kernels/activations_test.cc:306:16: error: variable length array used 
 T output_data[size];
              ^
/tensorflow/lite/micro/kernels/activations_test.cc:786:20: note: in instantiation of function template specialization 'tflite::testing::(anonymous namespace)::TestHardSwishQuantizedBias<unsigned char>' requested here
 tflite::testing::TestHardSwishQuantizedBias<uint8_t>(
                  ^
/tensorflow/lite/micro/kernels/activations_test.cc:307:25: error: variable length array used 
 T input_data_quantized[size];
                       ^
/tensorflow/lite/micro/kernels/activations_test.cc:324:27: error: variable length array used 
 float dequantized_output[output_elements_count];
`

**Comment by gbaned** [comment]
@mansnils Can you please check @rthadur's comments and resolve conflicts?. Thanks!

**Comment by mansnils** [comment]
@rthadur errors and conflics are resolved.

**Review by njeffrie — APPROVED** [review]

File: tensorflow/lite/micro/kernels/activations.cc [hunk 1]
```diff
@@ -77,6 +78,157 @@ inline void Relu6Quantized(Q lower, Q upper, const RuntimeShape& input_shape,
   }
 }
 
+inline std::int32_t RoundingDivideByPOT(std::int32_t numerator, int exponent) {
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations.cc:81]: Please avoid inlining code from tensorflow/lite/kernels/internal/reference/reference_ops.h.

Instead, take methods you want to share and create a new header, like we have done for other operators.  See conv.h for an example.  Add the new header to reference_ops.h and depend only on reference/hard_swish.h or reference/activations.h in the TFMicro kernel.

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 2]
```diff
@@ -135,6 +137,312 @@ void TestRelu6Float(const int* input_dims_data, const float* input_data,
   }
 }
 
+void GenerateUniformRandomVector(int size, float min, float max,
+                                 std::minstd_rand* random_engine,
+                                 std::vector<float>* result) {
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:142]: std::vector uses malloc under the hood.  Avoid on TF Micro.  Follow the example in other tests by passing in static size input / output arrays.

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 3]
```diff
@@ -135,6 +137,312 @@ void TestRelu6Float(const int* input_dims_data, const float* input_data,
...
+  std::vector<float> float_input_values;
+  for (int i = quantized_input_negative_three;
+       i <= quantized_input_positive_three; i++) {
+    float_input_values.push_back(
+        input_min +
+        (i - std::numeric_limits<T>::min()) * input_scale);
+  }
+  TF_LITE_MICRO_EXPECT_EQ(float_input_values.size(), size);
+
+  std::vector<float> float_ref_output_values;
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:296]: "

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 4]
```diff
@@ -135,6 +137,312 @@ void TestRelu6Float(const int* input_dims_data, const float* input_data,
...
+  TF_LITE_MICRO_EXPECT_LE(std::abs(bias), tolerated_bias);
+}
+
+void TestHardSwishFloat(const int size, float* output_data, std::minstd_rand* random_engine) {
+  std::vector<float> float_input_values;
+  const float kMin = -10.0f;
+  const float kMax = 10.0f;
+  GenerateUniformRandomVector(size, kMin, kMax, random_engine,
+                              &float_input_values);
+  std::vector<float> float_ref_output_values;
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:384]: "

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 5]
```diff
@@ -431,6 +739,79 @@ TF_LITE_MICRO_TEST(SimpleRelu6TestFloat) {
...
+  constexpr int size = 100;
+  float output_data[size] = {0.f};
+
+  tflite::testing::TestHardSwishFloat(size, output_data, &random_engine);
+}
+
+
+TF_LITE_MICRO_TEST(SimpleHardSwishTestInt8) {
+  std::minstd_rand random_engine;
+  std::vector<std::pair<float, float>> minmax_pairs{
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:753]: "

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 6]
```diff
@@ -431,6 +739,79 @@ TF_LITE_MICRO_TEST(SimpleRelu6TestFloat) {
...
+                                                      input_min, input_max, output_min, output_max,
+                                                      &random_engine);
+
+    }
+  }
+}
+
+TF_LITE_MICRO_TEST(SimpleHardSwishTestUint8) {
+  std::minstd_rand random_engine;
+  std::vector<std::pair<float, float>> minmax_pairs{
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:777]: "

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 7]
```diff
@@ -135,6 +137,312 @@ void TestRelu6Float(const int* input_dims_data, const float* input_data,
...
+    // It may actually be thanks to rounding, as std::minstd_rand::modulus
+    // is 2^31 - 1 is greater than the inverse float epsilon.
+    float random_value_scaled_0_1 =
+        (*random_engine)() *
+        (1.0f / static_cast<float>(std::minstd_rand::modulus));
+    (*result)[i] = min + (max - min) * random_value_scaled_0_1;
+  }
+}
+
+void EvalTestReferenceHardSwish(int size, const std::vector<float>& input,
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:162]: Please avoid std::vector here and below.

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 8]
```diff
@@ -135,6 +137,312 @@ void TestRelu6Float(const int* input_dims_data, const float* input_data,
...
+  // to allow for an off-by-one rounding error.
+  const float kTolerance = std::max(input_max - input_min, output_max - output_min) * (1.5f / 256.f);
+
+  TfLiteIntArray* input_dims = IntArrayFromInts(input_dims_data);
+  TfLiteIntArray* output_dims = IntArrayFromInts(output_dims_data);
+  const int output_elements_count = ElementCount(*output_dims);
+
+  TF_LITE_MICRO_EXPECT_EQ(output_elements_count, size);
+
+  std::vector<float> float_input_values;
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:194]: "

File: tensorflow/lite/micro/kernels/activations_test.cc [hunk 9]
```diff
@@ -135,6 +137,312 @@ void TestRelu6Float(const int* input_dims_data, const float* input_data,
...
+  // values.
+  TF_LITE_MICRO_EXPECT_LE(input_min, -3.0f);
+  TF_LITE_MICRO_EXPECT_GE(input_max, 3.0f);
+  const int quantized_input_negative_three =
+      std::round(std::numeric_limits<T>::min() +
+                 (-3.0f - input_min) / input_scale);
+  const int quantized_input_positive_three =
+      std::round(std::numeric_limits<T>::min() +
+                 (3.0f - input_min) / input_scale);
+  std::vector<float> float_input_values;
```
Comment (njeffrie) [inline tensorflow/lite/micro/kernels/activations_test.cc:287]: "

**Review by njeffrie — CHANGES_REQUESTED** [review]

**Comment by njeffrie** [comment]
Sorry for reversing approval.  I think I approved too early before.  There are a bunch of changes we will need to make to land this in TF Micro.

File: tensorflow/lite/micro/kernels/activations.cc [hunk 10]
```diff
@@ -148,13 +300,43 @@ TfLiteStatus Relu6Eval(TfLiteContext* context, TfLiteNode* node) {
...
-      TF_LITE_KERNEL_LOG(context, "Only float32 is supported currently, got %s",
+      TF_LITE_KERNEL_LOG(context, "Only float32/int8/uint8 are supported currently, got %s",
                          TfLiteTypeGetName(input->type));
       return kTfLiteError;
     }
   }
 }
 
+TfLiteStatus HardSwishPrepare(TfLiteContext* context, TfLiteNode* node) {
+  return kTfLiteOk;
```
[the remark this answers is not in the document]
Comment (mansnils) [reply tensorflow/lite/micro/kernels/activations.cc:304]: Added some checks and pre-calculating parameters in prepare.

File: tensorflow/lite/micro/kernels/activations.cc (hunk 1 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/activations.cc]: Moving RoundingDivideByPOT to new reference/activations.h. I thought TFLu should avoid using gemmlowp, that is why I added RoundingDivideByPOT.

File: tensorflow/lite/micro/kernels/activations_test.cc (hunk 2 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/activations_test.cc]: Done

File: tensorflow/lite/micro/kernels/activations_test.cc (hunk 7 above)
Comment (mansnils) [reply tensorflow/lite/micro/kernels/activations_test.cc]: All occurrences of std::vector are replaced.

**Comment by mansnils** [comment]
@njeffrie Thanks for the comments. Changes are now updated.

**Review by rthadur — DISMISSED** [review]

**Comment by rthadur** [comment]
@mansnils sorry for the delay , still seeing below errors , can you please check 

`tensorflow/lite/micro/kernels/activations_test.cc:407:51: error: implicit conversion turns string literal into bool: 'const char [14]' to 'bool' 
     CreateFloatTensor(output_data, output_dims, "output_tensor"),


tensorflow/lite/micro/kernels/activations_test.cc:406:57: error: implicit conversion turns string literal into bool: 'const char [13]' to 'bool' 
     CreateFloatTensor(float_input_values, input_dims, "input_tensor"),


tensorflow/lite/micro/kernels/activations_test.cc:212:48: error: implicit conversion turns string literal into bool: 'const char [14]' to 'bool' 
                           output_zero_point, "output_tensor"),


tensorflow/lite/micro/kernels/activations_test.cc:778:24: note: in instantiation of function template specialization 'tflite::testing::(anonymous namespace)::TestHardSwishQuantized<signed char>' requested here
     tflite::testing::TestHardSwishQuantized<int8_t>(
                      ^

tensorflow/lite/micro/kernels/activations_test.cc:210:29: error: implicit conversion turns string literal into bool: 'const char [13]' to 'bool' 
                           "input_tensor"),
                           ^~~~~~~~~~~~~~
tensorflow/lite/micro/kernels/activations_test.cc:212:48: error: implicit conversion turns string literal into bool: 'const char [14]' to 'bool' 
                           output_zero_point, "output_tensor"),
                                              ^~~~~~~~~~~~~~~
tensorflow/lite/micro/kernels/activations_test.cc:805:24: note: in instantiation of function template specialization 'tflite::testing::(anonymous namespace)::TestHardSwishQuantized<unsigned char>' requested here
     tflite::testing::TestHardSwishQuantized<uint8_t>(
                      ^

tensorflow/lite/micro/kernels/activations_test.cc:210:29: error: implicit conversion turns string literal into bool: 'const char [13]' to 'bool' 
                           "input_tensor"),
                           ^~~~~~~~~~~~~~

tensorflow/lite/micro/kernels/activations_test.cc:329:48: error: implicit conversion turns string literal into bool: 'const char [14]' to 'bool' 
                           output_zero_point, "output_tensor"),
                                              ^~~~~~~~~~~~~~~

tensorflow/lite/micro/kernels/activations_test.cc:826:20: note: in instantiation of function template specialization 'tflite::testing::(anonymous namespace)::TestHardSwishQuantizedBias<unsigned char>' requested here
 tflite::testing::TestHardSwishQuantizedBias<uint8_t>(
                  ^

tensorflow/lite/micro/kernels/activations_test.cc:327:29: error: implicit conversion turns string literal into bool: 'const char [13]' to 'bool' 
                           "input_tensor"),
`

**Comment by mansnils** [comment]
@rthadur Errors and conflicts resolved. Please re-approve.

**Review by rthadur — APPROVED** [review]
thank you

**Pull request merged**